### PR TITLE
SERVER-31378: Improve error checking of $dateToString

### PR DESCRIPTION
**Pull request by i80and**

### CODE REVIEW

File: src/mongo/db/query/datetime/date_time_support.h [hunk 1]
```diff
@@ -292,8 +286,12 @@ class TimeZone {
     void insertPadded(OutputStream& os, int number, int width) const {
         invariant(width >= 1);
         invariant(width <= 4);
-        invariant(number >= 0);
-        invariant(number <= 9999);
+
+        uassert(18537,
+                str::stream() << "$dateToString is only defined on date components 0-9999,"
```
Comment (cswanson310) [inline src/mongo/db/query/datetime/date_time_support.h]: While you're here, I believe there are plans to use this code for other things than $dateToString at some point (started but not committed as part of [SERVER=29627](https://jira.mongodb.org/browse/SERVER-29627)) Can you update this to instead be `"Could not convert date to string, date component was outside supported range of 0-9999: " << number` or something like that?

File: src/mongo/db/query/datetime/date_time_support_test.cpp [hunk 2]
```diff
@@ -37,6 +37,7 @@ namespace mongo {
 namespace {
 
 const TimeZoneDatabase kDefaultTimeZoneDatabase{};
+const char* kFormats[3] = {"%Y", "%G", NULL};
```
[the remark this answers is not in the document]
Comment (i80and) [reply src/mongo/db/query/datetime/date_time_support_test.cpp]: Updated!

My justification was to make it easier to make sure *all* format flags working with years are tested (which might prevent this sort of bug in the future)

File: src/mongo/db/query/datetime/date_time_support.h (hunk 1 above)
Comment (i80and) [reply src/mongo/db/query/datetime/date_time_support.h]: Done!

File: src/mongo/db/query/datetime/date_time_support_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (cswanson310) [reply src/mongo/db/query/datetime/date_time_support_test.cpp]: Gotcha. I understand the motivation, but there's nothing here to ensure an engineer who adds a formatter would remember to update this list, so until there are more than a few, or until there's a way to enumerate them, I'd prefer to keep them separate.

**Comment by cswanson310** [comment]
Looks good! I'll run this through evergreen, then squash+rebase+push!

**Comment by cswanson310** [comment]
Sorry for the delay - patch building here: https://evergreen.mongodb.com/version/59dce7f4e3c331590200b8bd

I hope to push this later today if all goes well.

**Pull request closed without merging**